Framework/DataObjects/src/TimeSplitter.cpp

```diff
@@ -37,6 +37,37 @@ TimeSplitter::TimeSplitter(const DateAndTime &start, const DateAndTime &stop) {
   clearAndReplace(start, stop, DEFAULT_VALUE);
 }
 
+/**
+ * Note: The amount of X values in input MatrixWorkspace must be 1 larger than the amount of Y values.
+ * There are NO undefined split regions here.
+ **/
+TimeSplitter::TimeSplitter(const Mantid::API::MatrixWorkspace_sptr &ws, const DateAndTime &offset) {
+  if (ws->getNumberHistograms() != 1) {
+    throw std::runtime_error("MatrixWorkspace can only have 1 histogram when constructing TimeSplitter.");
+  }
+
+  const auto X = ws->binEdges(0);
+  const auto &Y = ws->y(0);
+  if (std::any_of(X.begin(), X.end(), [](double i) { return static_cast<int>(i) < 0; })) {
+    throw std::runtime_error("All X values in MatrixWorkspace must be >= 0 to construct TimeSplitter.");
+  }
+  if (X.size() != Y.size() + 1) {
+    throw std::runtime_error(
+        "Size of x values must be one more than size of y values to construct TimeSplitter from MatrixWorkspace.");
+  }
+
+  int64_t offset_ns{offset.totalNanoseconds()};
+  for (size_t i = 1; i < X.size(); i++) {
+    auto timeStart = Types::Core::DateAndTime(X[i - 1], 0.0) + offset_ns;
+    auto timeEnd = Types::Core::DateAndTime(X[i], 0.0) + offset_ns;
+    auto index = static_cast<int>(Y[i - 1]);
+    if ((index != IGNORE_VALUE) && (valueAtTime(timeStart) != IGNORE_VALUE || valueAtTime(timeEnd) != IGNORE_VALUE)) {
+      g_log.warning() << "Workspace row " << i - 1 << " may be overwritten in conversion to TimeSplitter" << '\n';
+    }
```

Comment:
Since in the `MatrixWorkspace` case, we don't have rows, the warning will be a little more tricky. I think you might just say that the time interval between `<< timeStart << " and " << timeStop <<` may be overwritten.

Reply:
OK, I have to correct my comment a bit. We do have rows, but a row is like the whole (X,Y,E) array, not an individual array element!